mftb.md

```diff
@@ -0,0 +1,345 @@
+  without needing to dynamically import the code of token admin contracts.
+- **Forgo Deployment**: it should be possible to create custom token types without the deployment of
+  a custom fungible token contract. Token types are just data after all; token-type-specific
+  contracts––at least for basic functionality––are unnecessary.
+- **Extensible**: the `FungibleTokenbase` contract should offer all functionality necessary for a
+  3rd-party contract to become the token admin and craft use-case-specific functionality.
+- **Scalability**: the contract should support concurrent transactions at scale.
+- **Off-chain State Management API**: the contract-accompanying service should facilitate
+  interaction with persisted off-chain state (computing new merkle roots, paginating lists of tokens
+  and token accounts, reading allocation lists by allocation ID, allocator ID, allocated ID, token
+  ID, misc).
+
+### Non-goals
+
+- **Fine-grained Authorization**: there is no separation of administrative role by action. The token
+  admin can dispatch all actions. Meanwhile non-admin accounts can manipulate their token-specific
+  accounts (unless frozen by the token admin). To implement more advanced authorization, developers
+  can create and set as admin a contract that delegates back to the token base.
+- **Account Purging or "Sufficiency"**: the fungible token base does not represent existential
+  deposits nor weight custom tokens against native tokens such as Mina. There is no minimum balance.
+  When a token account has no funds, it is no longer represented in state. If it "receives" funds
+  (either from a transfer or from an allocation) it exists.
+
+### Actions
+
+It seems that actions and reducers are the recommended path for new o1js contracts, as this enables
+multiple actions to be queued in a single block (concurrent transactions (ish)). These actions are
+used to reduce a new state whenever a `Commit` action is dispatched. This means that users can run
+many commands in rapid succession while still ensuring inclusion later. This lazy state model is
+seemingly key to contract scalability. Therefore, the `FungibleTokenbase` contract API is described
+via action interfaces.
+
+> Note: the resulting state change descriptions are imprecise, as it is still unclear what
+> representations should live off-chain. For instance, the contract depends on representations of
+> potentially large mappings. Therefore, many of the actions will be accompanied by merkle paths and
+> proofs.
+
+> Note: errors are not yet represented.
+
+> Note: the actions + reducers approach has [several drawbacks](#actions-and-reducers-drawbacks)
+> which may impact the feasibility of a scalable `FungibleTokenbase` implementation.
+
+#### `Create`
+
+Our first action is to create a new token type. Everyone is authorized to dispatch this action.
+
+```ts
+interface Create {
+  /** A unique ID of TBD type––perhaps a `UInt64`. */
+  id: TokenId
```

Comment:
Is there a way to do so without breaking commutativity?

Reply:
idk how it breaks commutativity. tokenID can be just a hash of some unique struct

Reply:
How would this ID be generated within the contract? I'm not opposed to it, I just don't know how it could be implemented. I've contemplated tracking a `tokenCount: UInt64` in state and returning the current count from every `create` call using the pending o1js `@method.returns` feature. Unfortunately, this would break commutativity / might cause the same nonce precondition errors we saw in the initial non-action/reducer approach. Ideas on how we might generate a token ID without breaking commutativity?

Reply:
hash of smth